storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -95,13 +108,86 @@ func (s *commitLogSource) Read(
 	defer iter.Close()
 
 	var (
-		unmerged    = make(map[uint32]map[ts.Hash]encoderMap)
+		numShards = len(shardsTimeRanges)
+		unmerged  = make([]map[ts.Hash]encoderMap, numShards, numShards)
+		// TODO: Add to opts
+		numConc     = uint32(4)
 		bopts       = s.opts.ResultOptions()
 		blopts      = bopts.DatabaseBlockOptions()
 		blockSize   = bopts.RetentionOptions().BlockSize()
 		encoderPool = bopts.DatabaseBlockOptions().EncoderPool()
-		errs        = 0
 	)
+
+	encoderChans := []chan encoderArg{}
+	for i := uint32(0); i < numConc; i++ {
+		encoderChans = append(encoderChans, make(chan encoderArg, 1000))
+	}
+
+	// Spin up numConc background go-routines to handle M3TSZ encoding. This must
+	// happen before we start reading to prevent infinitely blocking writes to
+	// the encoderChans.
+	wg := sync.WaitGroup{}
+	workerErrs := make([]int, numConc, numConc)
+	for workerNum, encoderChan := range encoderChans {
+		wg.Add(1)
+		go func(workerNum int, ec <-chan encoderArg) {
+			for arg := range ec {
+				var (
+					series     = arg.series
+					dp         = arg.dp
+					unit       = arg.unit
+					annotation = arg.annotation
+					blockStart = arg.blockStart
+				)
+
+				unmergedShard := unmerged[series.Shard]
+				if unmergedShard == nil {
+					unmergedShard = make(map[ts.Hash]encoderMap)
+					unmerged[series.Shard] = unmergedShard
+				}
+
+				unmergedSeries, ok := unmergedShard[series.ID.Hash()]
+				if !ok {
+					unmergedSeries = encoderMap{
+						id:       series.ID,
+						encoders: make(map[int64][]encoder)}
+					unmergedShard[series.ID.Hash()] = unmergedSeries
+				}
+
+				var (
+					err           error
+					unmergedBlock = unmergedSeries.encoders[blockStart.UnixNano()]
+					wroteExisting = false
+				)
+				for i := range unmergedBlock {
+					if unmergedBlock[i].lastWriteAt.Before(dp.Timestamp) {
+						unmergedBlock[i].lastWriteAt = dp.Timestamp
+						err = unmergedBlock[i].enc.Encode(dp, unit, annotation)
+						wroteExisting = true
+						break
+					}
+				}
+				if !wroteExisting {
+					enc := encoderPool.Get()
+					enc.Reset(blockStart, blopts.DatabaseBlockAllocSize())
+
+					err = enc.Encode(dp, unit, annotation)
+					if err == nil {
+						unmergedBlock = append(unmergedBlock, encoder{
+							lastWriteAt: dp.Timestamp,
+							enc:         enc,
+						})
+						unmergedSeries.encoders[blockStart.UnixNano()] = unmergedBlock
+					}
+				}
+				if err != nil {
+					workerErrs[workerNum]++
+				}
+			}
+			wg.Done()
+		}(workerNum, encoderChan)
```

Comment:
Are you sure this is wrong? The reference you linked seems to suggest that what I'm doing is correct. If I didn't pass them into the goroutine it would be broken, but since I pass them in I'm creating a new variable that will hold onto the old values even as the for loop continues on.